from backend.api_gateway.libs.milkyhoop_prisma import Prisma
import asyncio
import signal
import logging
import os

import grpc
from grpc import aio
from grpc_health.v1 import health, health_pb2, health_pb2_grpc

from app.config import settings
from app import complaint_service_pb2_grpc as pb_grpc
from app import complaint_service_pb2 as pb

# âœ… Jika Prisma dipakai:
from app.prisma_client import prisma

# âœ… Logging config
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s"
)
logger = logging.getLogger(settings.SERVICE_NAME)

# âœ… gRPC handler implementasi
class Complaint_serviceServicer(pb_grpc.Complaint_serviceServicer):
    async def DoSomething(self, request, context):
        logger.info("ðŸ“¥ DoSomething request received: %s", request.input)
        return pb.Complaint_serviceResponse(
            status="ok",
            result=f"Processed input: {request.input}"
        )

    async def CreateComplaint(self, request, context):
        logger.info("ðŸ“¥ CreateComplaint received from user_id=%s: %s", request.user_id, request.message)

        # Simulasi penyimpanan data
        complaint_id = "C-" + os.urandom(4).hex()

        return pb.CreateComplaintResponse(
            status="success",
            complaint_id=complaint_id,
            message="Complaint received and logged."
        )

async def serve() -> None:
    if "DATABASE_URL" in os.environ:
        logger.info("ðŸ”Œ Connecting to Prisma...")
        await prisma.connect()
        logger.info("âœ… Prisma connected")

    server = aio.server()
    pb_grpc.add_Complaint_serviceServicer_to_server(Complaint_serviceServicer(), server)

    health_servicer = health.HealthServicer()
    health_pb2_grpc.add_HealthServicer_to_server(health_servicer, server)
    health_servicer.set('', health_pb2.HealthCheckResponse.SERVING)

    listen_addr = f"[::]:{settings.GRPC_PORT}"
    server.add_insecure_port(listen_addr)
    logger.info(f"ðŸš€ {settings.SERVICE_NAME} gRPC server listening on port {settings.GRPC_PORT}")

    stop_event = asyncio.Event()

    def handle_shutdown(*_):
        logger.info("ðŸ›‘ Shutdown signal received. Cleaning up...")
        stop_event.set()

    signal.signal(signal.SIGINT, handle_shutdown)
    signal.signal(signal.SIGTERM, handle_shutdown)

    try:
        await server.start()
        await stop_event.wait()
    finally:
        logger.info("ðŸ§¹ Shutting down gRPC server...")
        await server.stop(5)
        if "DATABASE_URL" in os.environ:
            logger.info("ðŸ§¹ Disconnecting Prisma...")
            await prisma.disconnect()
            logger.info("âœ… Prisma disconnected")
        logger.info("âœ… gRPC server shut down cleanly.")

if __name__ == "__main__":
    asyncio.run(serve())
